### PR TITLE
[Networking] Isolates DHT capability within ANs and ENs

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
The changes in this PR are as follows: 
1. It isolates the DHT capability within ANs and ENs. Currently, ever Flow node also runs as a DHT node, however, except for the ANs and ENs, the DHT capability is rendered redundant. ANs utilize the DHT as a routing discovery in public (unstaked) nodes to help unstaked ANs, followers, and observer nodes find each other in the pubsub. Moreover, ANs and ENs run DHT nodes to form a blob exchange service. Except for these two functionalities that are limited to ANs and ENs, running DHT nodes by other roles is a redundant overhead that also exposes them to DHT-based routing attacks. 

2. It removes dead codes associated with `Relay` and `Splitter` networks.

### CODE REVIEW

File: network/p2p/p2pnode/libp2pNode.go [hunk 1]
```diff
@@ -532,7 +532,7 @@ func (n *Node) IsConnected(peerID peer.ID) (bool, error) {
 // SetRouting may be called at most once.
 func (n *Node) SetRouting(r routing.Routing) {
 	if n.routing != nil {
-		n.logger.Fatal().Msg("routing already set")
+		irrecoverable.Throw(context.TODO(), fmt.Errorf("routing already set"))
```
Comment (peterargue) [inline network/p2p/p2pnode/libp2pNode.go:535]: I think it's better to stick with the fatal error unless there's a irrecoverable context to use.

File: network/p2p/p2pbuilder/libp2pNodeBuilder.go [hunk 2]
```diff
@@ -318,40 +296,44 @@ func (builder *LibP2PNodeBuilder) Build() (p2p.LibP2PNode, error) {
...
+			func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+				if builder.routingFactory != nil {
+					routingSystem, err := builder.routingFactory(ctx, h)
+					if err != nil {
+						ctx.Throw(fmt.Errorf("could not create routing system: %w", err))
+					}
+					node.SetRouting(routingSystem)
+					builder.gossipSubBuilder.SetRoutingSystem(routingSystem)
+				} else {
+					builder.logger.Warn().Msg("no routing system is set for the node; ignore the log for non-access and non-execution nodes")
```
Comment (peterargue) [inline network/p2p/p2pbuilder/libp2pNodeBuilder.go:309]: is this needed? I don't think we should log a warning on every startup for normal behavior. a missing routing system is a software bug and will have noticeable side effects during development.

File: network/p2p/p2pbuilder/libp2pNodeBuilder.go [hunk 3]
```diff
@@ -502,8 +484,21 @@ func DefaultNodeBuilder(
...
+
+		if r == flow.RoleAccess || r == flow.RoleExecution {
+			// Only access and execution nodes need to run DHT;
+			// Access nodes and execution nodes need DHT to run a blob service.
+			// Moreover, access nodes run a DHT to let un-staked (public) access nodes find each other on the public network.
+			builder.SetRoutingSystem(
+				func(ctx context.Context, host host.Host) (routing.Routing, error) {
+					return dht.NewDHT(ctx, host, protocols.FlowDHTProtocolID(sporkId), logger, metricsCfg.Metrics, dht.AsServer())
+				})
+		}
```
Comment (peterargue) [inline network/p2p/p2pbuilder/libp2pNodeBuilder.go:501]: should this go in scaffold? I feel like node specific behavior doesn't belong in the `DefaultNodeBuilder` method.

**Review by peterargue — APPROVED** [review]
Thanks for the cleanup! I guess these were left over from before we split the networks. Added a couple small suggestions, but otherwise looks good.

File: network/p2p/p2pnode/libp2pNode.go (hunk 1 above)
Comment (yhassanzadeh13) [reply network/p2p/p2pnode/libp2pNode.go]: https://github.com/onflow/flow-go/pull/4800/commits/bb1aec2cf82d7a63eba50da45118e3fc46972a46

File: network/p2p/p2pbuilder/libp2pNodeBuilder.go (hunk 2 above)
Comment (yhassanzadeh13) [reply network/p2p/p2pbuilder/libp2pNodeBuilder.go]: https://github.com/onflow/flow-go/pull/4800/commits/21e8489022d1377aa6fe058029dec4170227d218

File: network/p2p/p2pbuilder/libp2pNodeBuilder.go (hunk 3 above)
Comment (yhassanzadeh13) [reply network/p2p/p2pbuilder/libp2pNodeBuilder.go]: https://github.com/onflow/flow-go/pull/4800/commits/d0086b9d73ddf688a921e6c0c226404ca6aee41f

File: network/p2p/p2pbuilder/libp2pNodeBuilder.go [hunk 4]
```diff
@@ -44,6 +44,13 @@ import (
 	"github.com/onflow/flow-go/network/p2p/utils"
 )
 
+type RoutingSystemActivation bool
+
+const (
+	RoutingSystemActivationEnabled  RoutingSystemActivation = true
+	RoutingSystemActivationDisabled RoutingSystemActivation = false
```
Comment (kc1116) [inline network/p2p/p2pbuilder/libp2pNodeBuilder.go:51]: Would it be clearer to rename these to something that denotes the DHT server will be started. I.e `DhtServerEnabled`

File: network/p2p/p2pnode/libp2pNode.go [hunk 5]
```diff
@@ -532,6 +532,7 @@ func (n *Node) IsConnected(peerID peer.ID) (bool, error) {
 // SetRouting may be called at most once.
 func (n *Node) SetRouting(r routing.Routing) {
 	if n.routing != nil {
+		// we should not allow overriding the routing implementation if one is already set; crashing the node.
 		n.logger.Fatal().Msg("routing already set")
```
Comment (kc1116) [inline network/p2p/p2pnode/libp2pNode.go:535]: Wondering if we should start returning the error to the caller and allow the error to be thrown by the parents context?

**Review by kc1116 — APPROVED** [review]
Nice cleanup , left a couple optional suggestions.

File: network/p2p/p2pbuilder/libp2pNodeBuilder.go (hunk 4 above)
Comment (yhassanzadeh13) [reply network/p2p/p2pbuilder/libp2pNodeBuilder.go]: https://github.com/onflow/flow-go/pull/4800/commits/6247470cbf04fbeabf9abc062ce695e598d80b2b

File: network/p2p/p2pnode/libp2pNode.go (hunk 5 above)
Comment (yhassanzadeh13) [reply network/p2p/p2pnode/libp2pNode.go:535]: https://github.com/onflow/flow-go/pull/4800/commits/9311c0c466e1f201806f1c06f5bd034fcbb738cd

**Pull request merged**